File: 55.跳跃游戏.java
/*
 * @lc app=leetcode.cn id=55 lang=java
 *
 * [55] 跳跃游戏
 */

// @lc code=start
class Solution {
    public boolean canJump(int[] nums) {
        final int n = nums.length;
        int furthest = 0;
        for (int i = 0; i < n - 1; ++i) {
            furthest = Math.max(furthest, i + nums[i]);
            if (furthest <= i)  return false;
        }
        return furthest >= n - 1;
    }
}
// @lc code=end

File: 55.跳跃游戏.py
#
# @lc app=leetcode.cn id=55 lang=python3
#
# [55] 跳跃游戏
#
from typing import List

# @lc code=start
class Solution:
    def canJump(self, nums: List[int]) -> bool:
        """Solution: greedy"""
        n, rightmost = len(nums), 0
        for i in range(n):
            if i <= rightmost:   # the jump can reach at least ith location
                # if we have reached the ith location, we can then reach i+nums[i] location 
                rightmost = max(rightmost, i + nums[i])
                if rightmost >= n - 1:
                    return True
        return False


# @lc code=end
# s = Solution()
# print(s.canJump([3, 2, 1, 0, 4]))
